tkg/client/cluster.go

```diff
@@ -160,7 +160,13 @@ func (c *TkgClient) CreateCluster(options *CreateClusterOptions, waitForCluster
 			return false, errors.Wrap(err, "unable to get cluster configuration")
 		}
 
-		if !c.IsFeatureActivated(constants.FeatureFlagAllowLegacyCluster) {
+		allowLegacyCluster, err := c.TKGConfigReaderWriter().Get(constants.ConfigVariableAllowLegacyCluster)
+		if err != nil {
+			log.Error(err, "ALLOW_LEGACY_CLUSTER variable is set failed")
+			return false, err
```

Comment:
**question:** Is it catastrophic if we fail to get the ALLOW_LEGACY_CLUSTER env var? Especially if it isn't set. I think, not. Logging the fact is sufficient in my opinion. Then fallback on the feature flag.

Reply:
So, `err` is provably _always_ `nil`. Why bother checking it?

Reply:
@imikushin , no, `err` is not always `nil`. Because `c.TKGConfigReaderWriter().Get()` method will return a string and an error. like below:
```
// Get returns a configuration value of type string.
// In case the configuration value does not exists, it returns an error.
Get(key string) (string, error)
```
So check err is nil or not is necessary.

Reply:
See my other comment: https://github.com/vmware-tanzu/tanzu-framework/pull/4176#discussion_r1053911842

Reply:
@imikushin , error check is needed, because `c.TKGConfigReaderWriter().Get()` return a string and an error. if we want to below func returns a bool type,
```
func (c *TkgClient) allowLegacyCluster() (bool, error)
```
then we need convert the string value is returned by `c.TKGConfigReaderWriter().Get()` to a bool type use [strconv.ParseBool()](https://pkg.go.dev/strconv#ParseBool). Also `strconv.ParseBool` still returns a bool and an error. We can hidden error if happens.

Reply:
> error check is needed, because `c.TKGConfigReaderWriter().Get()` return a string and an error. 

The error it returns is solely to indicate that the variable is not set. This information can be used (in our decision logic) and error safely ignored (we could log the fact that the env var is not set): the var _not being set by the user_ is not an error condition.

> if we want to below func returns a bool type,
> 
> ```
> func (c *TkgClient) allowLegacyCluster() (bool, error)
> ```
> 
> then we need convert the string value is returned by `c.TKGConfigReaderWriter().Get()` to a bool type use [strconv.ParseBool()](https://pkg.go.dev/strconv#ParseBool). Also `strconv.ParseBool` still returns a bool and an error. We can't hidden error if happens.

If we want to fail cluster creation with an error, if _the user made a mistake_ setting an env var, that is a legitimate option.

But then, we shouldn't modify the env var, because if we do, we're confusing ourselves (whether the value in the env var is input from the user or our interpretation of user preferences set elsewhere), making the code needlessly more complex.

I'm basically saying, we don't need the method `SetAllowLegacyClusterConfiguration()`, and we should use `allowLegacyCluster()`, which should fully encapsulate the decision logic (of whether to allow legacy clusters), instead.

In general, if we can make a decision _without_ mutating state (_especially_ globally accessible state, like an env var), we should do it _without_ mutating state.

Reply:
The same thinking applies everywhere: 
- If the user provides input via an env var, we _should not_ modify this env var, but only read from it.
- If it is not absolutely necessary to mutate state to do something, definitely _do not mutate state_.

Reply:
@imikushin , `SetAllowLegacyClusterConfiguration` func is needed, because this func needs to do below actions:
- if user provides the env var, then we will only read it;
- if user doesn't provide the env var, then we will set it according to the feature gate's value (`features.cluster.allow-legacy-cluster`).

`SetAllowLegacyClusterConfiguration` is called during get related cluster configurations, namely `getClusterConfigurationBytes` func. We will generate legacy or classy cluster configurations according to func's return.
After the cluster configuration is generated, we call `allowLegacyCluster()` func again just to get `ALLOW_LEGACY_CLUSTER` env's value to decide if we want to use a converted config file to create cluster or not. We don't modify it's value.
```
bytes, err = c.getClusterConfigurationBytes(&options.ClusterConfigOptions, infraProviderName, isManagementCluster)
if err != nil {
     return false, errors.Wrap(err, "unable to get cluster configuration")
}

allowLegacyCluster, err := c.allowLegacyCluster()
if err != nil {
     return false, err
}
if !allowLegacyCluster {
     ..........
 }
```
In a word, `SetAllowLegacyClusterConfiguration()` will set `ALLOW_LEGACY_CLUSTER`'s value. `allowLegacyCluster()` only read `ALLOW_LEGACY_CLUSTER`'s value.

Reply:
```golang
bytes, err = c.getClusterConfigurationBytes(&options.ClusterConfigOptions, infraProviderName, isManagementCluster)
if err != nil {
     return false, errors.Wrap(err, "unable to get cluster configuration")
}

allowLegacyCluster, err := c.allowLegacyCluster() <======== (Here, I agree with you, we can ignore error.)
if err != nil {
     return false, err
}
if !allowLegacyCluster {
     ..........
 }
```
Because if error happens here, that means some network error or api client error, like you said, we can't do everything.

Reply:
> `SetAllowLegacyClusterConfiguration` func is needed, because this func needs to do below actions:
> 
> * if user provides the env var, then we will only read it;
> * if user doesn't provide the env var, then we will set it according to the feature gate's value (`features.cluster.allow-legacy-cluster`).

Yes, I can confirm it does these things. I'm also saying we don't really need it.

> `SetAllowLegacyClusterConfiguration` is called during get related cluster configurations, namely `getClusterConfigurationBytes` func. We will generate legacy or classy cluster configurations according to func's return. After the cluster configuration is generated, we call `allowLegacyCluster()` func again just to get `ALLOW_LEGACY_CLUSTER` env's value to decide if we want to use a converted config file to create cluster or not. We don't modify it's value.

Okay, I'm looking at the only place where it is used and I see this:

```go
	allowLegacyClusterCreated = c.SetAllowLegacyClusterConfiguration()
	if allowLegacyClusterCreated == "false" {
```

So, basically, its responsibility is to decide if legacy clusters are allowed (beside setting an env var). But don't we already have a good method for that? Yes we do, it is `allowLegacyCluster()`. Now, imagine `SetAllowLegacyClusterConfiguration()` didn't exist. We would then need to put all the logic from `SetAllowLegacyClusterConfiguration()` into `allowLegacyCluster()`, and call `allowLegacyCluster()` instead of `SetAllowLegacyClusterConfiguration()`.

We would end up setting an env var (if it is not set) and then reading from it. We'd naturally want to replace it with using a local variable, wouldn't we?

Reply:
@imikushin as we discussed in [Slack](https://vmware.slack.com/archives/C03UACASCFL/p1671428434106079), we may focus on an incorrect path. The `ALLOW_LEGACY_CLUSTER` variable should be a config variable rather than an environment variable, although it can be overridden via set an env var. 
So I still think we need to set it according to the feature gate's value if users don't config it in their cluster config file. Otherwise, we will read it from their config file.
How do you think?